### ISSUE TITLE
Pass something from http middleware to rpc middleware

### ISSUE THREAD

**Issue by dpc** [paritytech/jsonrpsee#1370]
I might have missed something, but I don't see a way to pass anything from http middlware to rpc middleware/handler.

I'd like to do authorization over http request, and then depending on the result possibly each rpc would enable/disable certain things etc.

Conceptually it's easy, but there seem to be nothing the two layers together, so it can't be done.

**Comment by niklasad1** [comment]
Hey hey,

Have you looked at [this example](https://github.com/paritytech/jsonrpsee/blob/master/examples/examples/jsonrpsee_as_service.rs) which does passes the HTTP headers to the RPC middleware stuff.

But you are right that it won't work with server builder API instead you have to utilize the "low-level API" and use jsonrpsee as as tower service and start the hyper server yourself.

Another example how we are [using it in substrate](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/client/rpc-servers/src/lib.rs#L170-#L239) to disable/enable rpc rate limiting middleware...

I hope that helps...

**Comment by dpc** [comment]
Thanks!

This is a bit mindbending. Do I understand correctly that the approach here is to create an rpc service for every http connection, kinda?

**Comment by niklasad1** [comment]
> This is a bit mindbending. Do I understand correctly that the approach here is to create an rpc service for every http connection, kinda?

Yes, if you want some state from HTTP request itself such as headers then it's only possible to utilize it by creating a rpc service per connection.

So I get that it's quite tricky to use/understand but that's the trade-off right because we didn't want to introduce any additional custom middleware stuff in jsonrpsee and clone stuff that may not used by users....

**Comment by dinhani** [comment]
You can pass data from HTTP to RPC with request extensions.

In this example we parse client identification in the [HTTP layer](https://github.com/cloudwalk/stratus/blob/acb53b202066c6324b43756810b2f484702bb9b6/src/eth/rpc/rpc_http_middleware.rs#L37-L43) and use it in the [RPC layer](https://github.com/cloudwalk/stratus/blob/acb53b202066c6324b43756810b2f484702bb9b6/src/eth/rpc/rpc_middleware.rs#L79).
